fs.go

```diff
@@ -17,96 +17,521 @@
+	s.listObjectMapMutex.Lock()
+	defer s.listObjectMapMutex.Unlock()
+
+	if walkChs, ok := s.listObjectMap[params]; ok {
+		for i, walkCh := range walkChs {
+			if !walkCh.timedOut {
+				newWalkChs := walkChs[i+1:]
+				if len(newWalkChs) > 0 {
+					s.listObjectMap[params] = newWalkChs
+				} else {
+					delete(s.listObjectMap, params)
+				}
+				return walkCh
+			}
+		}
+		// As all channels are timed out, delete the map entry
+		delete(s.listObjectMap, params)
+	}
+	return nil
+}
+
+// List of special prefixes for files, includes old and new ones.
+var specialPrefixes = []string{
+	"$multipart",
+	"$tmpobject",
+	"$tmpfile",
+	// Add new special prefixes if any used.
+}
+
+// List operation.
+func (s fsStorage) ListFiles(volume, prefix, marker string, recursive bool, count int) ([]FileInfo, bool, error) {
+	// Verify if volume is valid and it exists.
+	volumeDir, err := s.checkVolumeArg(volume)
+	if err != nil {
+		return nil, true, err
+	}
+	var fileInfos []FileInfo
+
+	if marker != "" {
+		// Verify if marker has prefix.
+		if marker != "" && !strings.HasPrefix(marker, prefix) {
+			return nil, true, errInvalidArgument
+		}
+	}
+
+	// Return empty response for a valid request when count is 0.
+	if count == 0 {
+		return nil, true, nil
+	}
+
+	// Over flowing count - reset to fsListLimit.
+	if count < 0 || count > fsListLimit {
+		count = fsListLimit
+	}
+
+	// Verify if prefix exists.
+	prefixDir := filepath.Dir(filepath.FromSlash(prefix))
+	prefixRootDir := filepath.Join(volumeDir, prefixDir)
+	if status, err := isDirExist(prefixRootDir); !status {
+		if err == nil {
+			// Prefix does not exist, not an error just respond empty list response.
+			return nil, true, nil
+		} else if strings.Contains(err.Error(), "not a directory") {
+			// Prefix exists as a file.
+			return nil, true, nil
+		}
+		// Rest errors should be treated as failure.
+		return nil, true, err
+	}
+
+	// Maximum 1000 files returned in a single call.
+	// Further calls will set right marker value to continue reading the rest of the files.
+	// popTreeWalker returns nil if the call to ListFiles is done for the first time.
+	// On further calls to ListFiles to retrive more files within the timeout period,
+	// popTreeWalker returns the channel from which rest of the objects can be retrieved.
+	walker := s.lookupTreeWalk(listParams{volume, recursive, marker, prefix})
+	if walker == nil {
+		walker = startTreeWalk(s.diskPath, volume, filepath.FromSlash(prefix), filepath.FromSlash(marker), recursive)
+	}
+	nextMarker := ""
+	for i := 0; i < count; {
+		walkResult, ok := <-walker.ch
+		if !ok {
+			// Closed channel.
+			return fileInfos, true, nil
+		}
+		// For any walk error return right away.
+		if walkResult.err != nil {
+			return nil, true, walkResult.err
+		}
+		fileInfo := walkResult.fileInfo
+		fileInfo.Name = filepath.ToSlash(fileInfo.Name)
+		// TODO: Find a proper place to skip these files.
+		// Skip temporary files.
```

Comment:
> Why can't we use https://github.com/minio/minio/blob/master/fs-dir-nix.go#L136?

This needs to be added, this is based off of earlier listObjects implementation.